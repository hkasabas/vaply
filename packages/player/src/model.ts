export type NodePlayerConfig = NodePackageConfig & {
  flow: NodeFlowConfig;
  nodes: Array<
    ContentNodeConfig & {
      annotations: Array<AnnotationConfig>;
      // triggers connecting content movement and annotations
      triggers: Array<AnnotationTriggerConfig>;
    }
  >;
};

export type NodePackageConfig = {
  id: string;
  title: string;
};

export type NodeFlowType = "list";
export type NodeFlowConfig = {
  type: NodeFlowType;
  // [<NodeFlowType>]: {},
  list: {
    nodes: string[];
  };
};

export type ContentNodeConfig = {
  code: string; // custom generated ID for internal referencing (eg. positioning)

  video: VideoContentConfig;
};

export type VideoContentType = "file";
export interface VideoContentConfig {
  type: VideoContentType;

  // [<VideoContentType>]: {},
  file?: FileVideoContentConfig;
}

export interface FileVideoContentConfig {
  src: string;
  poster?: string;
}

export type AnnotationType = "card" | "externalcontent" | "chapter";
export type AnnotationPositionType = "static" | "none";
export type AnnotationConfig = {
  code: string;
  type: AnnotationType;
  blocking?: boolean;
  dismissible?: boolean;

  position: {
    type: "static";

    static: Partial<{
      top: string; // CSS dimensions
      right: string; // CSS dimensions
      bottom: string; // CSS dimensions
      left: string; // CSS dimensions
    }>;
  };

  dimensions?: Partial<{
    width: string; // CSS dimensions
    height: string; // CSS dimensions
  }>;
} & (
  | { type: "card"; card: CardAnnotationConfig }
  | { type: "externalcontent"; externalcontent: ExternalContentAnnotationConfig }
  | { type: "chapter"; chapter: ChapterAnnotationConfig }
);

export type CardAnnotationConfig = {
  title?: string;
};
export type ExternalContentAnnotationConfig = {
  selector: string;
};
export type ChapterAnnotationConfig = {
  title: string;
};

export type AnnotationTriggerType = "timeupdate" | "route";
export type AnnotationTriggerConfig =
  | {
      targetCode: string; // annotation code
      type: "timeupdate";
      timeupdate: {
        start: string | number;
        end?: string | number;
      };
    }
  | {
      targetCode: string; // annotation code
      type: "route";
      route: {
        start: string | number;
      };
    };

// ---------- events

export type RouteEvent = string;

export type ObjectValue<T> = { value: T | undefined };
