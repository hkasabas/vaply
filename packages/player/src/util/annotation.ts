import { AnnotationConfig, AnnotationTriggerConfig, AnnotationTriggerType } from "@player/model";
import { assertUnreachable } from "@player/util/utils";

/** Typeguard fn that TS compiler understands when used in filters. */
function isNotEmpty<T>(obj: T): obj is NonNullable<T> {
  return obj != null;
}

export function findApplicableAnnotations(
  position: string,
  type: AnnotationTriggerType,
  availableAnnotations: AnnotationConfig[],
  triggers: AnnotationTriggerConfig[]
): AnnotationConfig[] {
  return (
    triggers
      // trigger by type
      .filter((trigger) => {
        return trigger.type === type;
      })
      // trigger by time
      .filter((trigger) => {
        return isActiveAnnotation(position, trigger);
      })
      // map trigger to annotation
      .map((trigger) => {
        return availableAnnotations.find((ann) => ann.code === trigger.targetCode);
      })
      // remove empty - triggers with targets not found in
      .filter(isNotEmpty)
  );
}

/** Remove previously closed annotations from given available annotation list. */
export function findActiveAnnotations(availableAnnotations: AnnotationConfig[], closedAnnotations: AnnotationConfig[]): AnnotationConfig[] {
  return (
    availableAnnotations
      // remove already closed
      .filter((cAnnot) => {
        return closedAnnotations.find((annot) => annot.code == cAnnot.code) == null;
      })
  );
}

/** Helper function which checks if there are any blocking annotations in given annotation list. */
export function hasBlockingAnnotations(availableAnnotations: AnnotationConfig[]): boolean {
  return (
    availableAnnotations
      // is there at least one that is blocking
      .some((annot) => annot.blocking)
  );
}

/** Return "start" specific for trigger type. */
export function findTriggerStart(trigger: AnnotationTriggerConfig): string | number | undefined {
  const type = trigger.type;
  switch (type) {
    case "route":
      return trigger.route.start;
    case "timeupdate":
      return trigger.timeupdate.start;
    default:
      assertUnreachable(type);
  }
}

// ----------

function isActiveAnnotation(position: string, trigger: AnnotationTriggerConfig): boolean {
  if (trigger.type === "timeupdate") {
    // in case of timeupdate annotations, position is number
    const timeNumber = Number.parseInt(position, 10);
    if (!Number.isNaN(timeNumber)) {
      // TODO: remove "as any"
      return (
        trigger.timeupdate != null &&
        (trigger.timeupdate.start as any) <= timeNumber &&
        (trigger.timeupdate.end == null || (trigger.timeupdate.end as any) > timeNumber)
      );
    }
    // NaN? no need to worry. This could be one of special routes that we just don't handle here yet
  }

  return false;
}
