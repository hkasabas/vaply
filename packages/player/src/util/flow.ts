import { NodeFlowConfig } from "@player/model";

export function getFirstNode(config: NodeFlowConfig): string | undefined {
  // default to "list" flow
  return getListFirstNode(config.list.nodes);
}

export function getPreviousNode(
  nodeId: string,
  config: NodeFlowConfig
): string | undefined {
  // default to "list" flow
  return getListPreviousNode(nodeId, config.list.nodes);
}

export function getNextNode(
  nodeId: string,
  config: NodeFlowConfig
): string | undefined {
  // default to "list" flow
  return getListNextNode(nodeId, config.list.nodes);
}

/**
 * List node flow
 *
 * List flow keeps nodes in a list and executes them sequentially.
 */

function getListFirstNode(nodes: string[]): string | undefined {
  return nodes.slice(0, 1).shift();
}

function getListPreviousNode(
  nodeId: string,
  nodes: string[]
): string | undefined {
  let previousId: string | undefined;
  let foundId: string | undefined;

  for (let i = 0; i < nodes.length; i++) {
    const id = nodes[i];
    if (id === nodeId) {
      foundId = previousId;
      break;
    }
    previousId = id;
  }

  return foundId;
}

function getListNextNode(nodeId: string, nodes: string[]): string | undefined {
  let previousId: string | undefined;
  let foundId: string | undefined;

  // we're looking for the next node so we'll iterate the arr in reverse to keep the same logic
  for (let i = nodes.length - 1; i >= 0; i--) {
    const id = nodes[i];
    if (id === nodeId) {
      foundId = previousId;
      break;
    }
    previousId = id;
  }

  return foundId;
}
