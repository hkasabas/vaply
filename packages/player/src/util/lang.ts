/**
 * Make certain type's properties optional.
 *
 * Example:
 * ```ts
 * interface A { a: number, b: number, c: number }
 *
 * type B = PartialBy<A, 'b' | 'c'>; // creates type B with all A's props except 'b' and 'c' are now optional
 * ```
 */
export type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

/**
 * Make all type's properties optional but also child properties, recursively.
 *
 * Example:
 * ```ts
 * interface A { a: { b: { c: number }}}
 *
 * type B = DeepPartial<A>; // creates type B with all propertis (a, b and c) optional
 * ```
 */
export type DeepPartial<T> = { [P in keyof T]?: DeepPartial<T[P]> };

/**
 * Make certain type's properties required.
 *
 * Example:
 * ```ts
 * interface A { a: number, b?: number, c?: number }
 *
 * type B = RequiredBy<A, 'b'>; // creates type B with 'a' and 'b' property required and 'c' property remains optional
 * ```
 */
export type RequiredBy<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>;

/**
 * Create new type by replacing type of certain properties.
 *
 * Example:
 * ```ts
 * interface Status { id: number; description: string; }
 * interface IdObj { id: number; }
 * interface A { a: number, b: { id: number, description: string }, c: { id: number, description: string } }
 *
 * type B = ReplaceBy<A, 'b' | 'c', IdObj>; // creates type B with all A's props except 'b' and 'c' are now of type IdObj
 * ```
 */
export type ReplaceBy<T, K extends keyof T, R> = Omit<T, K> & Record<K, R>;

/**
 * Optional type. Makes value of given type also both undefined and nullable.
 * Can be used for optional object properties but, unlike TS's integrated type "Partial", this can be used directly on object's property definitions.
 *
 * Example:
 * ```ts
 * class A {
 *   prop1: OptionalValue<string>;
 * }
 * ```
 *
 * is the same as
 * ```ts
 * class A {
 *   prop1: string | null | undefined;
 * }
 * ```
 */
export type OptionalValue<T> = T | undefined | null;

/**
 * Counter type to TS's integrated NonNullable. Obviously, it forces type to be null | undefined.
 * Can be used for TS type guards to distinguish between empty and non empty values.
 *
 * Example:
 * ```ts
 * function isEmpty<T>(value: T): value is Nullable<T> {
 *   return value == null;
 * }
 *
 * let value1: string | null | undefined;
 * if (isEmpty(value1)) {
 *   let value2 = value1; // value1 is ALWAYS null | undefined
 * } else {
 *   let value2 = value1; // value1 is NEVER null | undefined
 * }
 *
 * ```
 */
export type Nullable<T> = T extends null | undefined ? T : never;

/**
 * Utility type that complements TS's native `keyof` operator.
 * This operator returns new type consisting of all T's values.
 * This can be use to generate string literal "enumeration" types.
 *
 *
 * Example:
 * ```ts
 * // interfaces
 * interface Aaa {
 *   a: string;
 *   b: string;
 *   c: number;
 * }
 * type AaaValues = ValueOf(Aaa); // string | number
 *
 * // literals
 * const Bbb = {
 *   AAA: 'AAA',
 *   BBB: 'BBB',
 * } as const; // NOTE: cost assertion is important! Otherwise, TS compiler can only say that it's a string a not a string literal type;
 * type Bbb = ValueOf<typeof Bbb>; // 'AAA' | 'BBB'
 *
 * // extending literals
 * const Ccc = {
 *  ...Bbb, // copy BBB's values
 *  CCC: 'CCC'
 * } as const; // NOTE: const assertion is important
 * type CccValue = ValueOf<typeof Ccc>; // 'AAA' | 'BBB' | 'CCC'
 * const ccc1: CccValue = 'lalala'; // ERROR
 * const ccc2: CccValue = 'AAA'; // OK
 * const ccc3: CccValue = 'CCC'; // OK
 * ```
 */
export type ValueOf<T> = T[keyof T];
