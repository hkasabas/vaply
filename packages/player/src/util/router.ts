/**
 *
 * Node router
 *
 * Collection of methods for handling routing addresses
 *
 * There are two main structures:
 *  - address - string containing concatenated parts of route spearated by divider symbol
 *  - route - data structure with props for each distinct address part
 *
 * Id address doesn't contain divider, presume address is in format `NODE@POSITION`, otherwise presume `ADDRESS`
 */

const ADDRESS_DIVIDER_SYMBOL = '@';
const ANNOTATED_POSITION_SYMBOL = '#';

export type NodeRoute = {
  nodeCode?: string;
  position?: string;
};

// ----- special routes

/** Route that points to the start of targeted node */
export const NODE_ROUTE_NODE_START = 'NODE_START';
/** Route that points to the end of targeted node */
export const NODE_ROUTE_NODE_END = 'NODE_END';
/** Route that points to the next node */
export const NODE_ROUTE_NEXT_NODE = 'NEXT_NODE';
/** Route that points to the previous node */
export const NODE_ROUTE_PREVIOUS_NODE = 'PREVIOUS_NODE';

// ----- route/address

export function addressToRoute(address: string): NodeRoute {
  // if divider is present, presume address "NODE@POSITION", otherwise presum "ADDRESS"
  const parts = address.indexOf(ADDRESS_DIVIDER_SYMBOL) != -1 ? address.split(ADDRESS_DIVIDER_SYMBOL) : [undefined, address];

  return createRoute(...parts);
}

export function routeToAddress(route: NodeRoute): string {
  return createAddress(route.nodeCode, route.position);
}

export function createRoute(...parts: (string | undefined)[]): NodeRoute {
  return {
    nodeCode: parts[0],
    position: parts[1],
  };
}

export function createAddress(...parts: (string | undefined)[]): string {
  return parts.slice(0, 2).join(ADDRESS_DIVIDER_SYMBOL);
}

// ----- annotated positions

export function isAnnotatedPosition(position: string | undefined): boolean {
  return position != null ? position.startsWith(ANNOTATED_POSITION_SYMBOL) : false;
}

export function getAnnotatedPosition(position: string | undefined): string | undefined {
  return position != null ? position.substring(ANNOTATED_POSITION_SYMBOL.length) : undefined;
}
