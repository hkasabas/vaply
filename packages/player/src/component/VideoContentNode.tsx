import { FunctionComponent, JSX } from "preact";
import { useCallback, useEffect, useRef } from "preact/hooks";

import { ObjectValue, VideoContentConfig } from "@player/model";
import useUpdateEffect from "@player/util/hook/useUpdateEffect";
import { NODE_ROUTE_NEXT_NODE, NODE_ROUTE_NODE_END, NODE_ROUTE_NODE_START, NODE_ROUTE_PREVIOUS_NODE } from "@player/util/router";

/**
 * Video content node component props.
 */
export type VideoContentNodeProps = {
  config: VideoContentConfig;
  currentPosition?: ObjectValue<string>;
  blocked?: boolean;

  onReady?: () => void;
  onPlay?: () => void;
  onPause?: () => void;
  onEnd?: () => void;
  onTimeUpdate?: (time: number) => void;
};

/**
 * Video content node component
 *
 * Renders actual HTML video element, listens to it's events and sends them to content node.
 * It also updates video's current time on external navigation.
 */
const VideoContentNode: FunctionComponent<VideoContentNodeProps> = (props) => {
  const videElRef = useRef<HTMLVideoElement>(null);

  // ---------- props changes

  useUpdateEffect(() => {
    videElRef.current?.play();
  }, [props.config]);

  // blocked content
  useEffect(() => {
    if (videElRef.current != null) {
      if (props.blocked) {
        videElRef.current.pause();
      } else if (videElRef.current.paused && videElRef.current.currentTime !== 0) {
        videElRef.current.play();
      }
    }
  }, [props.blocked]);

  // route change
  useEffect(() => {
    if (videElRef.current != null && props.currentPosition?.value != null) {
      let newPosition;
      if (props.currentPosition.value == NODE_ROUTE_NODE_START) {
        console.log("props.currentPosition.value", props.currentPosition.value);
        newPosition = 0;
      } else if (props.currentPosition.value == NODE_ROUTE_NODE_END) {
        newPosition = videElRef.current.duration;
      } else if (props.currentPosition.value == NODE_ROUTE_PREVIOUS_NODE || props.currentPosition.value == NODE_ROUTE_NEXT_NODE) {
        newPosition = 0;
      } else {
        newPosition = Number.parseInt(props.currentPosition.value, 10);
      }

      videElRef.current.currentTime = newPosition;
    }
  }, [props.currentPosition]);

  // ---------- event handlers

  const handleCanPlay = useCallback(() => {
    props.onReady?.();
  }, [props.onReady]);

  const handlePlay = useCallback(() => {
    props.onPlay?.();
  }, [props.onPlay]);

  const handlePause = useCallback(() => {
    props.onPause?.();
  }, [props.onPause]);

  const handleEnd = useCallback(() => {
    props.onEnd?.();
  }, [props.onEnd]);

  const handleTimeUpdate = useCallback(
    (event: JSX.TargetedEvent<HTMLVideoElement>) => {
      props.onTimeUpdate?.(event.currentTarget.currentTime);
    },
    [props.onTimeUpdate]
  );

  const handleSeeked = useCallback(
    (event: JSX.TargetedEvent<HTMLVideoElement>) => {
      props.onTimeUpdate?.(event.currentTarget.currentTime);
    },
    [props.onTimeUpdate]
  );

  return (
    <div className="vaply-videoContentNode--contentContainer">
      {/* ----- file video ----- */}
      {props.config.type === "file" && (
        <video
          poster={props.config.file.poster}
          src={props.config.file.src}
          ref={videElRef}
          controls
          preload="metadata"
          loading="lazy"
          data-testid="video-el"
          onCanPlay={handleCanPlay}
          onPlay={handlePlay}
          onPause={handlePause}
          onEnded={handleEnd}
          onTimeUpdate={handleTimeUpdate}
          onSeeked={handleSeeked}
        />
      )}
    </div>
  );
};

export { VideoContentNode };
