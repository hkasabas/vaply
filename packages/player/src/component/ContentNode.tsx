import classNames from "classnames";
import { FunctionComponent } from "preact";
import { useCallback, useMemo, useState } from "preact/hooks";

import { VideoContentNode } from "@player/component/VideoContentNode";
import { Annotation } from "@player/component/annotation/Annotation";
import { ContentNodeConfig, RouteEvent, AnnotationConfig, AnnotationTriggerConfig, ObjectValue } from "@player/model";
import {
  findActiveAnnotations,
  findApplicableAnnotations as findAvailableAnnotations,
  findTriggerStart,
  hasBlockingAnnotations,
} from "@player/util/annotation";
import { getAnnotatedPosition, isAnnotatedPosition, NodeRoute, NODE_ROUTE_NEXT_NODE } from "@player/util/router";

/** Content node component props */
export type ContentNodeProps = {
  config: ContentNodeConfig;
  annotations: AnnotationConfig[];
  triggers: AnnotationTriggerConfig[];
  route?: NodeRoute;

  onRoute?: (route: RouteEvent) => void;
};

/**
 * Content node component
 *
 * Component in charge of all internal management inside a single node like
 * content starting, stopping, showing annotations, ...
 */
const ContentNode: FunctionComponent<ContentNodeProps> = (props) => {
  // track current time
  const [currentTime, setCurrentTime] = useState<number>(0);
  // track closed annotations
  const [closedAnnotations, setClosedAnnotations] = useState<AnnotationConfig[]>([]);

  // annotations available at the current moment
  const availableAnnotations = useMemo(() => {
    return findAvailableAnnotations(currentTime.toString(), "timeupdate", props.annotations, props.triggers);
  }, [currentTime]);

  // update resulting annotations
  const activeAnnotations = useMemo(() => {
    return findActiveAnnotations(availableAnnotations, closedAnnotations);
  }, [availableAnnotations, closedAnnotations]);

  // is content blocked iow. do we have an active blocking annotation
  const isContentBlocked = useMemo(() => {
    return hasBlockingAnnotations(activeAnnotations);
  }, [activeAnnotations]);

  // memoize CSS classes
  const containerClassNames = useMemo(() => {
    return classNames({
      "vaply-contentNode--contentContainer": true,
      "vaply-contentNode__blockingAnnotationMask": isContentBlocked ?? false,
    });
  }, [isContentBlocked]);

  // translate node route to content position
  const currentPosition: ObjectValue<string> | undefined = useMemo(() => {
    // annotated position
    if (isAnnotatedPosition(props.route?.position)) {
      const annotationName = getAnnotatedPosition(props.route?.position);
      return (
        props.triggers
          // find targeted annotation
          .filter((trigger) => trigger.targetCode === annotationName)
          // take it's start position
          .map((trigger) => ({ value: findTriggerStart(trigger)?.toString() }))
          .shift()
      );
    }
    // return route position as is
    return { value: props.route?.position };
  }, [props.route, props.triggers]);

  // ---------- handle events

  //  content timeupdate
  const handleContentTimeupdate = useCallback(
    (time: number) => {
      setCurrentTime(time);
    },
    [props.annotations, props.triggers]
  );

  // close annotation - Add annotation to closed annot list
  function handleAnnotationClose(targetAnnotation: AnnotationConfig) {
    setClosedAnnotations([
      ...closedAnnotations
        // prevent adding duplicate annotations
        .filter((annot) => annot.code !== targetAnnotation.code),
      targetAnnotation,
    ]);
  }

  // content node end
  const handleContentEnd = useCallback(() => {
    props.onRoute?.(NODE_ROUTE_NEXT_NODE);
  }, [props.onRoute]);

  return (
    <div className={containerClassNames}>
      {/* ---- Video node ---- */}
      {props.config.video && (
        <VideoContentNode
          currentPosition={currentPosition}
          config={props.config.video}
          blocked={isContentBlocked}
          onTimeUpdate={handleContentTimeupdate}
          onEnd={handleContentEnd}
        />
      )}

      {/* ----- Annotations ----- */}
      {activeAnnotations.map((annot) => (
        <Annotation key={annot.code} config={annot} onClose={() => handleAnnotationClose(annot)} />
      ))}
    </div>
  );
};

export { ContentNode };
