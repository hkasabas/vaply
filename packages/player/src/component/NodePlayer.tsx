import { h, FunctionComponent } from "preact";
import { useCallback, useEffect, useMemo, useState } from "preact/hooks";

import { ContentNode } from "@player/component/ContentNode";
import { NodeFlowConfig, NodePlayerConfig, ObjectValue } from "@player/model";
import { getFirstNode, getNextNode, getPreviousNode } from "@player/util/flow";
import { addressToRoute, createRoute, NodeRoute, NODE_ROUTE_NEXT_NODE, NODE_ROUTE_PREVIOUS_NODE } from "@player/util/router";

// TODO: import from path alias @player
import "../style/index.less";

/** Node player component props */
export interface NodePlayerProps {
  /**
   * Position content at specific node and position.
   *
   * This is an object to force triggering component update on repeated changes (which a simple string won't)
   */
  currentAddress?: ObjectValue<string>;

  /** Player node config */
  config: NodePlayerConfig;
}

/**Node player component
 *
 * This is the main entry component for a player. It comunicates with outside world,
 * resolves current node from address, displays it and handles all navigation between nodes.
 */
const NodePlayer: FunctionComponent<NodePlayerProps> = (props) => {
  const [currentAddress, setCurrentAddress] = useState<string | undefined>(props.currentAddress?.value);

  // this is the top component and we cannot pull this state upwards
  // so we need to use useEffect because this is syncing with "external system"
  useEffect(() => {
    setCurrentAddress(props.currentAddress?.value);
  }, [props.currentAddress?.value]);

  const currentRoute = useMemo(() => resolveRouteNode(currentAddress, props.config.flow), [currentAddress, props.config.flow]);
  // current node
  const currentNode = useMemo(
    () => (currentRoute != null ? props.config.nodes.find((n) => n.code === currentRoute.nodeCode) : undefined),
    [currentRoute, props.config]
  );

  const currentNodeAnnotations = useMemo(() => {
    const currentNode = currentRoute != null ? props.config.nodes.find((n) => n.code === currentRoute.nodeCode) : undefined;
    return currentNode?.annotations ?? [];
  }, [currentRoute, props.config.nodes]);

  const currentNodeTriggers = useMemo(() => {
    const currentNode = currentRoute != null ? props.config.nodes.find((n) => n.code === currentRoute.nodeCode) : undefined;
    return currentNode?.triggers ?? [];
  }, [currentRoute, props.config.nodes]);

  // ---------- event handlers

  // route event
  const handleNodeRoute = useCallback(
    (address: string | undefined) => {
      setCurrentAddress(address);
    },
    [setCurrentAddress]
  );

  return (
    <div className="vaply-nodePlayer__container">
      {currentNode && (
        <ContentNode config={currentNode} route={currentRoute} annotations={currentNodeAnnotations} triggers={currentNodeTriggers} onRoute={handleNodeRoute} />
      )}
    </div>
  );
};

function resolveRouteNode(address: string | undefined, flow: NodeFlowConfig): NodeRoute | undefined {
  let route: NodeRoute | undefined = address != null ? addressToRoute(address) : undefined;

  // empty address, fallback to the first node that flow returns
  if (route == null || route?.nodeCode == null) {
    route = route ?? {}; // initialize with empty obj if empty

    route = createRoute(getFirstNode(flow), route.position);

    if (route.nodeCode == null) throw "Cannot detect route node! Is the node flow empty?!";
  }

  if (route.position != null) {
    // next
    if (route.position === NODE_ROUTE_NEXT_NODE) {
      const nextNode = getNextNode(route.nodeCode, flow);
      if (nextNode != null) {
        // position is relative so ignore it in new route
        route = { nodeCode: nextNode };
      }
    }
    // previous
    else if (route.position === NODE_ROUTE_PREVIOUS_NODE) {
      const previousNode = getPreviousNode(route.nodeCode, flow);
      if (previousNode != null) {
        // position is relative so ignore it in new route
        route = { nodeCode: previousNode };
      }
    }
    // leave everything as is
  }

  return route;
}

export { NodePlayer };
