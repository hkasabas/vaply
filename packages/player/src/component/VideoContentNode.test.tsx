import { fireEvent, render, waitFor } from "@testing-library/preact";

import { VideoContentNode } from "@player/component/VideoContentNode";
import { VideoContentConfig } from "@player/model";
import { NODE_ROUTE_NEXT_NODE, NODE_ROUTE_NODE_END, NODE_ROUTE_NODE_START, NODE_ROUTE_PREVIOUS_NODE } from "@player/util/router";

describe("video content node", () => {
  const VIDEO: VideoContentConfig = {
    type: "file",
    file: { src: "./demo_video_longer.mp4" },
  };
  const MOCK_VIDEO_DURATION = 100;

  beforeAll(() => {
    mockVideo(MOCK_VIDEO_DURATION);
  });
  afterAll(() => {
    jest.restoreAllMocks();
  });

  it("renders", () => {
    // render
    const result1 = render(<VideoContentNode config={VIDEO} />);
    expect(result1.container).toMatchSnapshot();
  });

  it('should start/stop video when "blocked" property changes', async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    // "play" video and move time manually since it cannot really run
    video.play();
    video.currentTime = 10;

    result.rerender(<VideoContentNode config={VIDEO} blocked />);

    expect(video.paused).toBeTruthy();

    result.rerender(<VideoContentNode config={VIDEO} blocked={false} />);

    expect(video.paused).toBeFalsy();
  });

  it("should start video on config change", async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    expect(video.paused).toBeTruthy();

    // rerender with new config object
    result.rerender(<VideoContentNode config={{ ...VIDEO }} />);

    expect(video.paused).toBeFalsy();
  });

  it("should position video to start on NODE_START", async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    // initial time
    video.currentTime = 10;

    // set position to NODE_START
    result.rerender(<VideoContentNode config={VIDEO} currentPosition={{ value: NODE_ROUTE_NODE_START }} />);

    expect(video.currentTime).toBe(0);
  });

  it("should position video to start on NODE_END", async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    // initial time
    video.currentTime = 10;

    // set position to NODE_END
    result.rerender(<VideoContentNode config={VIDEO} currentPosition={{ value: NODE_ROUTE_NODE_END }} />);

    expect(video.currentTime).toBe(MOCK_VIDEO_DURATION);
  });

  it("should position video to the start on PREVIOUS_NODE", async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    // initial time
    video.currentTime = 10;

    // set position to PREVIOUS_NODE
    result.rerender(<VideoContentNode config={VIDEO} currentPosition={{ value: NODE_ROUTE_PREVIOUS_NODE }} />);

    expect(video.currentTime).toBe(0);
  });

  it("should position video to the start on NEXT_NODE", async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    // initial time
    video.currentTime = 10;

    // set position to NEXT_NODE
    result.rerender(<VideoContentNode config={VIDEO} currentPosition={{ value: NODE_ROUTE_NEXT_NODE }} />);

    expect(video.currentTime).toBe(0);
  });

  it("should position video to specified time", async () => {
    const result = render(<VideoContentNode config={VIDEO} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    // initial time
    video.currentTime = 10;

    // set position to NEXT_NODE
    result.rerender(<VideoContentNode config={VIDEO} currentPosition={{ value: "56" }} />);

    expect(video.currentTime).toBe(56);
  });

  it("should listen to events", async () => {
    const onReadyFn = jest.fn();
    const onPlayFn = jest.fn();
    const onPauseFn = jest.fn();
    const onEndFn = jest.fn();
    const onTimeUpdateFn = jest.fn();

    const result = render(
      <VideoContentNode config={VIDEO} onReady={onReadyFn} onPlay={onPlayFn} onPause={onPauseFn} onEnd={onEndFn} onTimeUpdate={onTimeUpdateFn} />
    );
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    fireEvent.canPlay(video);
    fireEvent.play(video);
    fireEvent.pause(video);
    fireEvent.ended(video);
    video.currentTime = 10;
    fireEvent.seeked(video);
    video.currentTime = 12;
    fireEvent.timeUpdate(video);

    expect(onReadyFn).toHaveBeenCalledTimes(1);
    expect(onPlayFn).toHaveBeenCalledTimes(1);
    expect(onPauseFn).toHaveBeenCalledTimes(1);
    expect(onEndFn).toHaveBeenCalledTimes(1);
    expect(onTimeUpdateFn).toHaveBeenCalledTimes(2);
    expect(onTimeUpdateFn.mock.calls).toEqual([[10], [12]]);
  });

  it('should listen to "play" event', async () => {
    const fn = jest.fn();

    const result = render(<VideoContentNode config={VIDEO} onPlay={fn} />);
    const video = result.getByTestId("video-el") as HTMLVideoElement;

    fireEvent.play(video);

    expect(fn).toHaveBeenCalledTimes(1);
  });
});

function mockVideo(duration: number) {
  // Mocking paused property
  Object.defineProperty(HTMLMediaElement.prototype, "paused", {
    writable: true,
    value: true, // Default paused state
  });

  // Mocking currentTime property
  Object.defineProperty(HTMLMediaElement.prototype, "currentTime", {
    writable: true,
    value: 0, // Default starting time
  });

  // Mocking duration property
  Object.defineProperty(HTMLMediaElement.prototype, "duration", {
    writable: true,
    value: duration, // Default duration
  });

  // Mocking play method
  HTMLMediaElement.prototype.play = jest.fn().mockImplementation(function (this: HTMLMediaElement) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (this as any).paused = false; // Update paused state
    return Promise.resolve(); // play() is asynchronous
  });

  // Mocking pause method
  HTMLMediaElement.prototype.pause = jest.fn().mockImplementation(function (this: HTMLMediaElement) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (this as any).paused = true; // Update paused state
  });
}
