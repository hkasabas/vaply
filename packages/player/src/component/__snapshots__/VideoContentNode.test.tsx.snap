// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`video content node renders 1`] = `
<div>
  <div
    class="vaply-videoContentNode--contentContainer"
  >
    <video
      controls=""
      data-testid="video-el"
      loading="lazy"
      preload="metadata"
      src="./demo_video_longer.mp4"
    />
  </div>
</div>
`;
